### DISCUSSION TITLE
Best practices for updating config

### DISCUSSION THREAD

**Discussion by itdependsnetworks** [mkdocs/mkdocs#3196] [Q&A]
What are the best practices to add to the config, things like `extra_css` or `markdown_extensions`? Based on github searches, I have come up with the below patterns.

``` python
    def on_config(self, config, **kwargs):
        if "markdown_extensions" not in config:
            config["markdown_extensions"] = []
        if "pymdownx.tilde" not in config['markdown_extensions']:
            config['markdown_extensions'].append('pymdownx.tilde')
        if "pymdownx.details" not in config['markdown_extensions']:
            config['markdown_extensions'].append('pymdownx.details')
        if "pymdownx.superfences" not in config['markdown_extensions']:
            config['markdown_extensions'].append('pymdownx.superfences')
        config['extra_css'].append(f'classy.css')
        return config

    def on_post_build(self, config, **kwargs):
        """Copy the css into where we defined it before."""
        src = os.path.abspath(os.path.join(os.path.dirname(__file__), './css/classy.css'))
        dst = os.path.join(config['site_dir'], 'classy.css')
        copy(src, dst)
```

I don't think that they are complete though, as en example, if you have these options configured in similar but not exactly the same way, there would be a conflict, e.g. 

``` yaml
  - pymdownx.superfences:
     custom_fences:
        - name: rich
          class: rich
          format: !!python/name:textual._doc.rich
```

I have tried searching multiple ways and reading the docs, such as the example shown here: https://www.mkdocs.org/dev-guide/plugins/#on_event_name but have not found an answer.

**Comment by kamilkrzyskow** [comment]
Hello 👋 , 
the `MkDocsConfig` separates extensions names and their configuration during load.

https://github.com/mkdocs/mkdocs/blob/1f1bfad76336573ec3811ddba943ca0e9cc4cccf/mkdocs/config/defaults.py#L95-L98
https://github.com/mkdocs/mkdocs/blob/1f1bfad76336573ec3811ddba943ca0e9cc4cccf/mkdocs/config/defaults.py#L100-L101
https://github.com/mkdocs/mkdocs/blob/1f1bfad76336573ec3811ddba943ca0e9cc4cccf/mkdocs/config/config_options.py#L838-L847

Some general practice advice:
- Consider using a loop to make the `on_config` a bit more readable. 
- Those options will always exist even if not present in the `mkdocs.yml`, so I would opt out of doing the initial `not in config` check and `[]` assignment. 
- For the path lookup of the file I would personally use `pathlib.Path` over the `os.path` in this case.

If that's the best way to copy the custom.css file, I'm not sure 🤔. Maybe adding the path to the `static_templates` or `theme.dirs` would be cleaner, but I'm also not sure if that would copy them over 😅 Food for thought, I guess.  

I've recently developed a hook, which modified `mdx_configs`, you can [check it out here](https://github.com/squidfunk/mkdocs-material/discussions/5365#discussioncomment-5617501) if you want some reference. However it isn't a 1:1 usage case like here.

**Reply by itdependsnetworks** [reply]
> Consider using a loop to make the on_config a bit more readable.

Makes sense, thanks.

> Those options will always exist even if not present in the mkdocs.yml, so I would opt out of doing the initial not in config check and [] assignment.

ok, will remove the starting two lines

> For the path lookup of the file I would personally use pathlib.Path over the os.path in this case.

ok, makes sense 

I end up with:

``` python
    def on_config(self, config, **kwargs):
        extensions = ["pymdownx.tilde", "pymdownx.details", "pymdownx.superfences"]
        for extension in extensions:
            if extension not in config['markdown_extensions']:
                config['markdown_extensions'].append(extension)
        config['extra_css'].append('classy.css')
        return config

    def on_post_build(self, config, **kwargs):
        """Copy the css into where we defined it before."""
        src = str(Path(__file__).resolve().parent), './css/classy.css'))
        dst = os.path.join(config['site_dir'], 'classy.css')
        copy(src, dst)
```

Will test out `static_templates` or `theme.dirs` over the next few days, thanks!

**Reply by kamilkrzyskow** [reply]
That's a plugin, right? Then store the path in `self` to avoid having multiple raw / "magic" strings.
That `copy` function is the `shutil.copy`, right? It should support Path-like objects too.
Keep the quotation marks consistent `black` formatter recommends using primarily `"`. 
```py
def __init__(self, *args, **kwargs):
    super().__init__(*args, **kwargs):
    self.classy_path = Path(__file__).parent / "css" / "classy.css"
    
def on_config(self, config, **__):
    # ...
    config["extra_css"].append(self.classy_path.name)
    
def on_post_build(self, config, **__):
    copy(self.classy_path, Path(config["site_dir"]) / self.classy_path.name)
```
